drivers/syslog/syslog_channel.c

```diff
@@ -234,23 +229,31 @@ g_syslog_channel[CONFIG_SYSLOG_MAX_CHANNELS] =
 #ifdef CONFIG_SYSLOG_DEFAULT
 static int syslog_default_putc(FAR syslog_channel_t *channel, int ch)
 {
-  UNUSED(channel);
-
 #  ifdef CONFIG_ARCH_LOWPUTC
+  /* See https://github.com/apache/nuttx/issues/14662
+   * about what this critical section is for.
+   */
+
+  irqstate_t flags = enter_critical_section();
   up_putc(ch);
+  leave_critical_section(flags);
```

Comment:
To be honest, I don't see any benefit to protecting `up_putc()`, unlike `up_nputs()` which guarantees single-line integrity.

Reply:
please read the description in https://github.com/apache/nuttx/issues/14662.
this is NOT about the "messages-intermixed" issue.

Reply:
I'm not sure if you've noticed that some drivers use `spinlock` + `irqsave` + `tx fifo polling` to implement `up_putc()`:
https://github.com/apache/nuttx/blob/master/drivers/serial/uart_pl011.c#L671-L679
For the current workaround, I think it might be more profitable to make similar changes in `up_putc()`, since we're not entirely sure if there's other code calling `up_putc()`.

Reply:
spin_lock_irqsave doesn't fix the issue because serial.c uses critical section, not spinlock.

Reply:
But whether it is the serial driver or syslog driver, up_putc() will eventually be called, right ?  TX interrupt will wake up serial driver, but the final output is still protected by up_putc(), right?

Reply:
Should serial.c use a spinlock / spin_lock_irqsave ? I cannot imagine why the big kernel lock (enter/exit_critical) needs to be invoked there (although I have not looked at the driver itself).

if CONFIG_SPINLOCK = 0, then spin_lock_irqsave reverts back to a local interrupt disable. Should work ?

Reply:
> But whether it is the serial driver or syslog driver, up_putc() will eventually be called, right ? TX interrupt will wake up serial driver, but the final output is still protected by up_putc(), right?

no. drivers/serial uses a separate api. (uart_ops_s)

Reply:
> Should serial.c use a spinlock / spin_lock_irqsave ? I cannot imagine why the big kernel lock (enter/exit_critical) needs to be invoked there (although I have not looked at the driver itself).

because we don't have a spinlock-interlocked blocking api? (do we?)

> if CONFIG_SPINLOCK = 0, then spin_lock_irqsave reverts back to a local interrupt disable. Should work ?

i'm not sure what you mean here. is it about this PR?

Reply:
> because we don't have a spinlock-interlocked blocking api? (do we?)
Not one that knows to yield. 

In SMP case using enter/exit_critical does exactly the same thing as a spin lock but it locks many unrelated (not serial related) APIs in the kernel (like semaphores etc.) as well, as they use enter/exit_critical. 

> i'm not sure what you mean here. is it about this PR?
Just that in non-SMP case using spin_lock_irqsave would revert back to a local interrupt disable, like it does when using enter/exit_critical (if someone is worried about non-SMP perf.)

I seem to remember seeing a discussion about serial/syslog locking and where it should be, but IMO the mutual exclusion (whatever the chosen mechanism is..) should be on the device level as the shared resource here is the serial port device? So replacing the critical section from both serial.c and syslog.c by a single spinlock should not be an issue. As it would be the serial port lock.

Also, doesn't syslog support other devices than serial port as well ? Are they protected by enter/exit_critcal  (I honestly don't know).